File: vbsrc/Samples/XAML_Features/AttachedProperties/SampleAttachedProperties.vb

#If SLMIGRATION
Imports System.Windows
#Else
using Windows.UI.Xaml;
#End If

Namespace Global.OpenSilver.Samples.VBShowcase
    Public Module SampleAttachedProperties
        Public Function GetAriaLabel(ByVal obj As DependencyObject) As String
            Return CStr(obj.GetValue(AriaLabelProperty))
        End Function

        Public Sub SetAriaLabel(ByVal obj As DependencyObject, ByVal value As String)
            obj.SetValue(AriaLabelProperty, value)
        End Sub

        Public ReadOnly AriaLabelProperty As DependencyProperty = DependencyProperty.RegisterAttached(name:="AriaLabel", propertyType:=GetType(String), ownerType:=GetType(SampleAttachedProperties), defaultMetadata:=New PropertyMetadata(defaultValue:=Nothing) With {
.MethodToUpdateDom = AddressOf AriaLabel_MethodToUpdateDom
})

        ''' <summary>
        ''' This method is called when the DOM tree is rendered.
        ''' </summary>
        ''' <paramname="d">The dependency object to which the property is attached</param>
        ''' <paramname="newValue">The new value of the attached property</param>
        Public Sub AriaLabel_MethodToUpdateDom(ByVal d As DependencyObject, ByVal newValue As Object)
            If TypeOf d Is FrameworkElement Then
                ' Convert the value to string:
                Dim value As String = (If(newValue IsNot Nothing, newValue.ToString(), ""))

                ' Get a reference to the <div> DOM element used to render the UI element:
                Dim div = Interop.GetDiv(CType(d, FrameworkElement))

                ' Set the "Aria Label" attribute on the <div> via a JavaScript interop call:
                Interop.ExecuteJavaScript("$0.setAttribute('aria-label', $1)", div, value)

                'Note: for documentation related to the commands above, please refer to https://opensilver.net/links/how-to-call-javascript.aspx
            End If
        End Sub

    End Module
End Namespace

File: src/Samples/XAML_Features/AttachedProperties/SampleAttachedProperties.fs
namespace OpenSilver.Samples.Showcase

open System
open System.Collections.Generic
open System.Linq
open System.Text
open System.Threading.Tasks
#if SLMIGRATION
open System.Windows
#else
open Windows.UI.Xaml
#endif

type SampleAttachedProperties() =
    static member GetAriaLabel (obj : DependencyObject) =
        obj.GetValue(SampleAttachedProperties.AriaLabelProperty) :?> string

    static member SetAriaLabel(obj : DependencyObject, value : string) =
        obj.SetValue(SampleAttachedProperties.AriaLabelProperty, value)

    static member AriaLabel_MethodToUpdateDom (d : DependencyObject) (newValue : obj) =
        if d :? FrameworkElement then
            let value = if newValue <> null then newValue.ToString() else ""

            // Get a reference to the <div> DOM element used to render the UI element:
            let div = OpenSilver.Interop.GetDiv(d :?> FrameworkElement)

            // Set the "Aria Label" attribute on the <div> via a JavaScript interop call:
            OpenSilver.Interop.ExecuteJavaScript("$0.setAttribute('aria-label', $1)", div, value) |> ignore
            // Note: For documentation related to the commands above, please refer to https://opensilver.net/links/how-to-call-javascript.aspx

    static member AriaLabelProperty =
        DependencyProperty.RegisterAttached("AriaLabel", typeof<string>, typeof<SampleAttachedProperties>,
            PropertyMetadata(defaultValue = null, MethodToUpdateDom = SampleAttachedProperties.AriaLabel_MethodToUpdateDom))
